amazon/ion/core.py

```diff
@@ -374,28 +406,85 @@ class Timestamp(datetime):
         if TIMESTAMP_FRACTION_PRECISION_FIELD in kwargs:
             fractional_precision = kwargs.get(TIMESTAMP_FRACTION_PRECISION_FIELD)
-            if fractional_precision is not None and 1 > fractional_precision > MICROSECOND_PRECISION:
+            if fractional_precision is not None and fractional_precision < 1:
                 raise ValueError('Cannot construct a Timestamp with fractional precision of %d digits, '
-                                 'which is out of the supported range of [1, %d].'
-                                 % (fractional_precision, MICROSECOND_PRECISION,))
+                                 'which is less than 1.' % fractional_precision)
             # Make sure we mask this before we construct the datetime.
             del kwargs[TIMESTAMP_FRACTION_PRECISION_FIELD]
+        if TIMESTAMP_FRACTIONAL_SECONDS_FIELD in kwargs:
+            fractional_seconds = kwargs.get(TIMESTAMP_FRACTIONAL_SECONDS_FIELD)
+            # Fractional seconds should be a Decimal in the range [0, 1), or None.
+            if fractional_seconds is not None and fractional_seconds > 1:
+                raise ValueError('Fractional seconds cannot be greater than 1.')
+            # Make sure we mask this before we construct the datetime.
+            del kwargs[TIMESTAMP_FRACTIONAL_SECONDS_FIELD]
+        if len(args) > 6:
+            if DATETIME_MICROSECONDS is not None and fractional_precision is not None \
+                    and fractional_seconds is not None:
+                if scale_to_precision(DATETIME_MICROSECONDS, fractional_precision) != fractional_seconds:
+                    raise ValueError('Microseconds and fractional seconds are not equivalent.')
+
+            if fractional_seconds is not None and fractional_precision is None:
+                raise ValueError('Fractional precision cannot be None while fractional seconds is not None.')
+
+            if DATETIME_MICROSECONDS is not None and fractional_precision is not None \
+                    and fractional_seconds is None:
+                if fractional_precision > 6:
+                    fractional_seconds = scale_to_precision(DATETIME_MICROSECONDS, fractional_precision)
+                    lst = list(args)
+                    integral_fractional_seconds = fractional_seconds * (10 ** MICROSECOND_PRECISION)
+                    lst[MICROSECOND_ARGUMENT_INDEX] = Decimal(integral_fractional_seconds).quantize(
+                        MICROSECOND_PRECISION,
+                        rounding="ROUND_DOWN")
+                    args = tuple(lst)
+                else:
+                    fractional_seconds = scale_to_precision(DATETIME_MICROSECONDS, MICROSECOND_PRECISION)
+
+            if DATETIME_MICROSECONDS is None and fractional_seconds is not None:
+                lst = list(args)
+                integral_fractional_seconds = fractional_seconds * (10 ** MICROSECOND_PRECISION)
+                lst[MICROSECOND_ARGUMENT_INDEX] = int(integral_fractional_seconds)
```

Comment:
There is some "don't repeat yourself" going on here, factor out this rounding into its own function, the wall of conditionals here is really hard to follow and that would help it.

Reply:
Done, factored out into its own function.